README.md
# libSSE-php

[![License](https://img.shields.io/badge/License-MIT-428F7E.svg)](LICENSE.md)
[![Build Status](https://travis-ci.org/tonyhhyip/libSSE-php.svg?branch=master)](https://travis-ci.org/tonyhhyip/libSSE-php)

An easy-to-use, object-oriented library for Server-Sent Events

## Installation

To install this package you'll need [composer](https://getcomposer.org/).

Run `composer require tonyhhyip/sse`


## Usage

Server-side(PHP):

```php
	<?php
	require_once('/path/to/vendor/autoload.php'); //Load with ClassLoader
	
	use Sse\Event;
	use Sse\SSE;
	
	//create the event handler
	class YourEventHandler implements Event {
		public function update(){
			//Here's the place to send data
			return 'Hello, world!';
		}
		
		public function check(){
			//Here's the place to check when the data needs update
			return true;
		}
	}
	
	$sse = new SSE(); //create a libSSE instance
	$sse->addEventListener('event_name', new YourEventHandler());//register your event handler
	$sse->start();//start the event loop
	?>
```

Client-side(javascript):
```javascript
	var sse = new EventSource('path/to/your/sse/script.php');
	sse.addEventListener('event_name',function(e){
		var data = e.data;
		//handle your data here
	},false);
```



## Settings

After you created the libSSE instance, there's some settings for you to control the behaviour.
Below is the settings provided by the library.

```php

<?php
	require_once('/path/to/vendor/autoload.php'); //Load with ClassLoader
	
	use Sse\SSE;
	
	$sse = new SSE();
    $sse->set($property, $value);
```


Direct access of property is kept with magic method for backward compatible.
```php
	<?php
	require_once('/path/to/vendor/autoload.php'); //Load with ClassLoader
	
	use Sse\SSE;
	
	$sse = new SSE();
	
	$sse->exec_limit = 10; //the execution time of the loop in seconds. Default: 600. Set to 0 to allow the script to run as long as possible.
	$sse->sleep_time = 1; //The time to sleep after the data has been sent in seconds. Default: 0.5.
	$sse->client_reconnect = 10; //the time for the client to reconnect after the connection has lost in seconds. Default: 1.
	$sse->use_chunked_encoding = true; //Use chunked encoding. Some server may get problems with this and it defaults to false
	$sse->keep_alive_time = 600; //The interval of sending a signal to keep the connection alive. Default: 300 seconds.
	$sse->allow_cors = true; //Allow cross-domain access? Default: false. If you want others to access this must set to true.
	?>
```

## Updates

1. Add Support of Symfony Http Foundation Compoent
2. SSE use magic method instead of direct access
3. Add Redis and Memcahce Mechnism
4. Add SessionLike Mechnism
5. Fixed event loop handling where removing handlers at runtime can result in a broken state.
6. Use Symfony HttpFoundation StreamedResponse to replace the old version
7. Add Changelog and contributing guide.

## Special For PHP 5.3 and 5.4
If you see and error message like `your PHP version does not satisfy that requirement.`,
please remove composer.lock and re-install it.

## Documentation

You may find it here.
[https://github.com/licson0729/libSSE-php/wiki/libSSE-docs](https://github.com/licson0729/libSSE-php/wiki/libSSE-docs)

## Development

This is an active project. If you want to help me please suggest ideas to me and track issues or find bugs. If you like it, please consider star it to let more people know.

## Compatibility

Because server-sent events is a new standard and still in flux, only certain browsers support it.
However, polyfill for server-sent events is available.
Also on shared hosting, it may disable PHP's `set_time_limit` function and the library may not work as excepted.
There's some settings in the library that can fix it.

## Integration with Frameworks

### Symfony
```php

    <?php
    use Symfony\Bundle\FrameworkBundle\Controller\Controller;
    use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
    use Sse\SSE;
    
    class DefaultController extends Controller
    {
        /**
         * @Route("/sse", name="sse")
         */
        public function sseAction()
        {
            $sse = new SSE();
            // Add your event listener
            return $sse->createResponse();
        }
    }
 ```   
### Laravel
Please use [laravel-sse](https://github.com/tonyhhyip/laravel-sse).


### Yii2
Please use [yii2-sse](https://github.com/odannyc/yii2-sse).

## Contribution
Please see the [CONTRIBUTING.md](CONTRIBUTING.md).


CHANGELOG.md
CHANGELOG
===========

## 2.2.0

 * Add PHP 8 attribute `ReturnTypeWillChange`
 * Fix `ob_flush` no buffer to flush issue

## 2.1.2

 * Use apcu functions via polyfill for PHP 5.3+ compat

## 2.1.1

 * Fix X-Accel-Buffering syntax.
 * Add doc string comments
 * Updates readme for better readability and adds installation instructions

## 2.1.0

 * Add Support of Symfony Http Foundation Compoent
 * Add PubSub Service
 * SSE use magic method instead of direct access
 * Add Redis and Memcahce Mechnism
 * Add SessionLike Mechnism
 * Fixed event loop handling where removing handlers at runtime can result in a broken state.
 
## 2.0.1

 * Fix Order of arguments due to change of Utils::sseBlock
 
## 2.0.0

 * Use namespaces directly
 * SSEEvent becomes interfaces \SSE\Event
 * rename SSEData as \SSE\Data
 * Add DataInterface for all Mechnism
 * SSEUtils rename as \SSE\Utils
 * All snake case method name rename to camel case
 * Change order of argument of Utils::sseBlock
  
## 1.0.0

 * Add to Packagist as tonyhhtip/sse


src/SSE.php
<?php
/**
 * libSSE-php
 *
 * Copyright (C) Licson Lee, Tony Yip 2016.
 *
 * Permission is hereby granted, free of charge,
 * to any person obtaining a copy of this software
 * and associated documentation files (the "Software"),
 * to deal in the Software without restriction,
 * including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons
 * to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice
 * shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS",
 * WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 * @category libSSE-php
 * @author   Licson Lee <licson0729@gmail.com>
 * @author   Tony Yip <tony@opensource.hk>
 * @license  http://opensource.org/licenses/MIT MIT License
 */

namespace Sse;

use ArrayAccess;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpFoundation\StreamedResponse;

class SSE implements ArrayAccess
{

    /**
     * @var array
     */
    private $handlers = array();

    /**
     * Event ID.
     *
     * @var int
     */
    private $id = 0;

    /**
     * @var int
     */
    private $start;

    /**
     * Config Setting
     * @var array
     */
    private $config = array(
        'sleep_time' => 0.5,                // seconds to sleep after the data has been sent
        'exec_limit' => 600,                // the time limit of the script in seconds
        'client_reconnect' => 1,            // the time client to reconnect after connection has lost in seconds
        'allow_cors' => false,              // Allow Cross-Origin Access?
        'keep_alive_time' => 300,           // The interval of sending a signal to keep the connection alive
        'is_reconnect' => false,            // A read-only flag indicates whether the user reconnects
        'use_chunked_encoding' => false,    // Allow chunked encoding
    );

    /**
     * SSE constructor.
     *
     * @param Request $request
     * @return void
     */
    public function __construct(Request $request = null)
    {
        //if the HTTP header 'Last-Event-ID' is set
        //then it's a reconnect from the client

        if ($request === null) {
            $request = Request::createFromGlobals();
        }

        $this->id = intval($request->headers->get('Last-Event-ID', 0));
        $this->config['is_reconnect'] = $request->headers->has('Last-Event-ID');

    }

    /**
     * Attach a event handler
     * @param string $event the event name
     * @param Event $handler the event handler
     * @return void
     */
    public function addEventListener($event, Event $handler)
    {
        $this->handlers[$event] = $handler;
    }

    /**
     * remove a event handler
     *
     * @param string $event the event name
     * @return void
     */
    public function removeEventListener($event)
    {
        unset($this->handlers[$event]);
    }

    /**
     * Get all the listeners
     *
     * @return array
     */
    public function getEventListeners()
    {
        return $this->handlers;
    }

    /**
     * Has listener
     * @return bool
     */
    public function hasEventListener()
    {
        return count($this->handlers) !== 0;
    }

    /**
     * Start the event loop
     *
     * @return null
     */
    public function start(){
        $response = $this->createResponse();
        $response->send();
    }

    /**
     * Send Data in buffer to client
     */
    public function flush()
    {
        @ob_flush();
        @flush();
    }

    /**
     * Send Data
     *
     * @param string $content
     */
    private function send($content)
    {
        print($content);
    }

    /**
     * Send a SSE data block
     *
     * @param mixed $id Event ID
     * @param string $data Event Data
     * @param string $name Event Name
     */
    public function sendBlock($id, $data, $name = null)
    {
        $this->send("id: {$id}\n");
        if (strlen($name) && $name !== null) {
            $this->send("event: {$name}\n");
        }

        $this->send($this->wrapData($data) . "\n\n");
    }

    /**
     * Create SSE data string
     *
     * @param string $string data to be processed
     * @return string
     */
    private function wrapData($string)
    {
        return 'data:' . str_replace("\n","\ndata: ", $string);
    }

    /**
     * Get time start
     * @return int
     */
    public function getUptime()
    {
        return time() - $this->start;
    }

    /**
     * Get the number tick
     * @return bool
     */
    public function isTick()
    {
        return $this->getUptime() % $this->keep_alive_time === 0;
    }

    /**
     * Sleep the process
     */
    public function sleep()
    {
        usleep($this->sleep_time * 1000000);
    }

    /**
     * Returns a Symfony HTTPFoundation StreamResponse.
     *
     * @return StreamedResponse
     */
    public function createResponse()
    {
        $this->init();
        $that = $this;
        $callback = function () use ($that) {
            $that->setStart(time());
            echo 'retry: ' . ($that->client_reconnect * 1000) . "\n";	// Set the retry interval for the client
            while (true) {
                // Leave the loop if there are no more handlers
                if (!$that->hasEventListener()) {
                    break;
                }

                if ($that->isTick()) {
                    // No updates needed, send a comment to keep the connection alive.
                    // From https://developer.mozilla.org/en-US/docs/Server-sent_events/Using_server-sent_events
                    echo ': ' . sha1(mt_rand()) . "\n\n";
                }
                
                // Start to check for updates
                foreach ($that->getEventListeners() as $event => $handler) {
                    if ($handler->check()) { // Check if the data is avaliable
                        $data = $handler->update(); // Get the data
                        $id = $that->getNewId();
                        $that->sendBlock($id, $data, $event);
                        
                        // Make sure the data has been sent to the client
                        $that->flush();
                    }
                }

                // Break if the time exceed the limit
                if ($that->exec_limit !== 0 && $that->getUptime() > $that->exec_limit) {
                    break;
                }
                // Sleep
                $that->sleep();
            }
        };


        $response = new StreamedResponse($callback, Response::HTTP_OK, array(
            'Content-Type' => 'text/event-stream',
            'Cache-Control' => 'no-cache',
            'X-Accel-Buffering' => 'no' // Disables FastCGI Buffering on Nginx
        ));

        if($this->allow_cors){
            $response->headers->set('Access-Control-Allow-Origin', '*');
            $response->headers->set('Access-Control-Allow-Credentials', 'true');
        }

        if($this->use_chunked_encoding)
            $response->headers->set('Transfer-encoding', 'chunked');

        return $response;
    }

    /**
     * Get the id for new message
     *
     * @return int
     */
    public function getNewId()
    {
        $this->id += 1;
        return $this->id;
    }

    /**
     * Initial System
     *
     * @return void
     */
    protected function init()
    {
        @set_time_limit(0); // Disable time limit

        // Prevent buffering
        if(function_exists('apache_setenv')){
            @apache_setenv('no-gzip', 1);
        }

        @ini_set('zlib.output_compression', 0);
        @ini_set('implicit_flush', 1);

        while (ob_get_level() != 0) {
            ob_end_flush();
        }
        ob_implicit_flush(1);
    }

    public function setStart($start)
    {
        $this->start = $start;
    }

    public function getStart()
    {
        return $this->start;
    }

    /**
     * Get config of SSE
     * @param string $key
     *
     * @return mixed
     */
    public function get($key)
    {
        return $this->config[$key];
    }

    /**
     * Get config of SSE
     *
     * @param string $key
     *
     * @return mixed
     */
    public function __get($key)
    {
        return $this->get($key);
    }

    /**
     * Set config of SSE
     * @param string $key
     * @param mixed $value
     * @return void
     */
    public function set($key, $value)
    {
        if (in_array($key, array('is_reconnect'))) {
            throw new \InvalidArgumentException('is_reconnected is an read-only flag');
        }
        $this->config[$key] = $value;
    }

    /**
     * Set config of SSE
     * @param string $key
     * @param mixed $value
     * @return void
     */
    public function __set($key, $value)
    {
        $this->set($key, $value);
    }

    /**
     * Determine if the given attribute exists.
     *
     * @param string $offset
     * @return bool
     */
    #[\ReturnTypeWillChange]
    public function offsetExists($offset)
    {
        return isset($this->config[$offset]);
    }

    /**
     * Get the value for a given offset.
     * @param string $offset
     * @return mixed
     */
    #[\ReturnTypeWillChange]
    public function offsetGet($offset)
    {
        return $this->get($offset);
    }

    /**
     * Set the value for a given offset.
     *
     * @param string $offset
     * @param mixed $value
     * @return void
     */
    #[\ReturnTypeWillChange]
    public function offsetSet($offset, $value)
    {
        $this->set($offset, $value);
    }

    /**
     * Unset the value for a given offset.
     *
     * @param  string $offset
     * @return void
     */
    #[\ReturnTypeWillChange]
    public function offsetUnset($offset)
    {
        $keys = array('sleep_time', 'exec_limit', 'client_reconnect', 'allow_cors', 'keep_alive_time', 'is_reconnect', 'use_chunked_encoding');
        if (in_array($offset, $keys)) {
            throw new \InvalidArgumentException($offset . ' is not allowed to removed');
        }

        unset($this->config[$offset]);
    }
}
